commit f6c0a63b7b9288e691e56b14fa9c88c8cd9253b6 (refs/changes/25/1425/6)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2017-03-01T14:19:24-08:00 (2 years, 7 months ago)
    
    MORAY-396 update to cueball v2
    Reviewed by: David Pacheco <dap@joyent.com>

diff --git a/CHANGES.md b/CHANGES.md
index f70d20a..905f8bc 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -1,5 +1,9 @@
 # Changelog
 
+## v3.1.1
+
+* [MORAY-396](http://smartos.org/bugview/MORAY-396) update to cueball v2
+
 ## v3.1.0
 
 * [MORAY-254](http://smartos.org/bugview/MORAY-254) moray tools need manual
diff --git a/lib/client.js b/lib/client.js
index 65ff0f6..f606290 100644
--- a/lib/client.js
+++ b/lib/client.js
@@ -134,7 +134,8 @@ function MorayClient(options) {
         'resolver': resolver,
         'recovery': cueballOptions.recovery,
         'target': cueballOptions.target,
-        'maximum': cueballOptions.maximum
+        'maximum': cueballOptions.maximum,
+        'connectionHandlesError': true
     });
 
     /* Internal state. */
diff --git a/lib/pool.js b/lib/pool.js
index e8bcf4c..b8203be 100644
--- a/lib/pool.js
+++ b/lib/pool.js
@@ -38,12 +38,14 @@ module.exports = MorayConnectionPool;
  * This is a struct-like class that represents a single logical connection.  The
  * lifecycle and logic are managed by the MorayConnectionPool class.
  */
-function MorayConnection(key, conn, log) {
+function MorayConnection(key, conn, hdl, log) {
     assert.string(key, 'key');
     assert.object(conn, 'conn');
+    assert.object(hdl, 'hdl');
     assert.object(log, 'log');
     this.mc_key = key;      /* cueball identifier for this connection */
     this.mc_conn = conn;    /* object implementing Cueball "Connection" */
+    this.mc_hdl = hdl;      /* cueball handle to release after drain */
     this.mc_log = log;      /* bunyan-style logger */
     this.mc_nreqs = 0;      /* number of outstanding requests */
 }
@@ -95,8 +97,8 @@ function MorayConnectionPool(args) {
     this.mcp_nalloc_fail = 0;   /* failed allocations */
     this.mcp_nreleased = 0;     /* releases */
 
-    this.mcp_cueball_set.on('added', function onConnectionAdd(key, conn) {
-        self.connAdd(key, conn);
+    this.mcp_cueball_set.on('added', function onConnectionAdd(key, conn, hdl) {
+        self.connAdd(key, conn, hdl);
     });
 
     this.mcp_cueball_set.on('removed', function onConnectionRemoved(key) {
@@ -169,7 +171,7 @@ MorayConnectionPool.prototype.connRelease = function (aconn) {
  * [private] Invoked by cueball when a new connection has been established and
  * is ready for use.  Just add it to our set of available connections.
  */
-MorayConnectionPool.prototype.connAdd = function connAdd(key, conn) {
+MorayConnectionPool.prototype.connAdd = function connAdd(key, conn, hdl) {
     var mconn, extras;
 
     assert.ok(!this.mcp_conns.hasOwnProperty(key));
@@ -177,7 +179,8 @@ MorayConnectionPool.prototype.connAdd = function connAdd(key, conn) {
 
     extras = conn.socketAddrs();
     extras.key = key;
-    mconn = new MorayConnection(key, conn, this.mcp_log.child(extras, true));
+    mconn = new MorayConnection(key, conn, hdl,
+        this.mcp_log.child(extras, true));
     this.mcp_conns[key] = mconn;
     this.mcp_avail[key] = true;
     mconn.mc_log.info('new connection');
@@ -227,5 +230,5 @@ MorayConnectionPool.prototype.connDelete = function (key) {
     delete (this.mcp_conns[key]);
 
     mconn.mc_log.info('removed connection');
-    mconn.mc_conn.destroy();
+    mconn.mc_hdl.release();
 };
diff --git a/package.json b/package.json
index 615e0a6..8ad1858 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
     "name": "moray",
     "description": "Moray client library",
-    "version": "3.1.0",
+    "version": "3.1.1",
     "author": "Joyent (joyent.com)",
     "keywords": [ "moray" ],
     "main": "./lib/index.js",
@@ -14,8 +14,8 @@
         "backoff": "^2.4.1",
         "bunyan": "^1.3.2",
         "cmdutil": "^1.1.0",
-        "cueball": "^1.0.3",
-	"extsprintf": "^1.3.0",
+        "cueball": "^2.2.0",
+        "extsprintf": "^1.3.0",
         "fast": "^2.1.0",
         "libuuid": "0.2.1",
         "jsprim": "^1.3.0",
